File: Eje_Semana_6/C++/SistemaReserva.cpp
#include <iostream>
#include <string>

class Reserva {
public:
    Reserva(std::string nombre_del_pasajero, std::string numero_de_vuelo, std::string fecha)
        : nombre_del_pasajero(nombre_del_pasajero), numero_de_vuelo(numero_de_vuelo), fecha(fecha) {}

    std::string mostrarDetalle() {
        return "Reserva para " + nombre_del_pasajero + "\nNúmero de vuelo: " + numero_de_vuelo + "\nFecha: " + fecha;
    }

protected:
    std::string nombre_del_pasajero;
    std::string numero_de_vuelo;
    std::string fecha;
};

class ReservaEconomica : public Reserva {
public:
    ReservaEconomica(std::string nombre_del_pasajero, std::string numero_de_vuelo, std::string fecha, std::string asiento)
        : Reserva(nombre_del_pasajero, numero_de_vuelo, fecha), asiento(asiento) {}

    std::string mostrarDetalle() {
        std::string padre = Reserva::mostrarDetalle();
        return padre + "\nClase: Económica\nAsiento: " + asiento;
    }

private:
    std::string asiento;
};

class ReservaBusiness : public Reserva {
public:
    ReservaBusiness(std::string nombre_del_pasajero, std::string numero_de_vuelo, std::string fecha, std::string asiento, std::string vip)
        : Reserva(nombre_del_pasajero, numero_de_vuelo, fecha), asiento(asiento), vip(vip) {}

    std::string mostrarDetalle() {
        std::string padre = Reserva::mostrarDetalle();
        return padre + "\nClase: Business\nAsiento: " + asiento + "\nAcceso a sala VIP: " + vip;
    }

private:
    std::string asiento;
    std::string vip;
};

class ReservaPrimeraClase : public Reserva {
public:
    ReservaPrimeraClase(std::string nombre_del_pasajero, std::string numero_de_vuelo, std::string fecha, std::string asiento, std::string primeraclase)
        : Reserva(nombre_del_pasajero, numero_de_vuelo, fecha), asiento(asiento), primeraclase(primeraclase) {}

    std::string mostrarDetalle() {
        std::string padre = Reserva::mostrarDetalle();
        return padre + "\nClase: Primera Clase\nAsiento: " + asiento + "\nServicio de Champagne: " + primeraclase;
    }

private:
    std::string asiento;
    std::string primeraclase;
};

int main() {
    ReservaEconomica reservaEconomica("Juan Pérez", "AA123", "2023-09-20", "Economica 23A");
    ReservaBusiness reservaBusiness("María García", "BA456", "2023-09-20", "vip 5B", "Sí");
    ReservaPrimeraClase reservaPrimeraClase("Pedro Rodríguez", "CX789", "2023-09-20", "Primera Clase 1A", "Sí");

    std::cout << "Detalle de la reserva económica:" << std::endl;
    std::cout << reservaEconomica.mostrarDetalle() << std::endl;

    std::cout << "\nDetalle de la reserva de business:" << std::endl;
    std::cout << reservaBusiness.mostrarDetalle() << std::endl;

    std::cout << "\nDetalle de la reserva de primera clase:" << std::endl;
    std::cout << reservaPrimeraClase.mostrarDetalle() << std::endl;

    return 0;
}

File: Eje_Semana_6/Python/SistemaReserva.py
class Reserva:
    def __init__(self, nombre_del_pasajero, numero_de_vuelo, fecha):
        self.nombre_del_pasajero = nombre_del_pasajero
        self.numero_de_vuelo = numero_de_vuelo
        self.fecha = fecha

    def mostrar_detalle(self):
        return f"Reserva para {self.nombre_del_pasajero}\nNúmero de vuelo: {self.numero_de_vuelo}\nFecha: {self.fecha}"

class ReservaEconomica(Reserva):
    def __init__(self, nombre_del_pasajero, numero_de_vuelo, fecha, asiento):
        super().__init__(nombre_del_pasajero, numero_de_vuelo, fecha)
        self.asiento = asiento

    def mostrar_detalle(self):
        padre = super().mostrar_detalle()
        return f"{padre}\nClase: Económica\nAsiento: {self.asiento}"

class ReservaBusiness(Reserva):
    def __init__(self, nombre_del_pasajero, numero_de_vuelo, fecha, asiento, vip):
        super().__init__(nombre_del_pasajero, numero_de_vuelo, fecha)
        self.asiento = asiento
        self.vip = vip

    def mostrar_detalle(self):
        padre = super().mostrar_detalle()
        return f"{padre}\nClase: Business\nAsiento: {self.asiento}\nAcceso a sala VIP: {self.vip}"

class ReservaPrimeraClase(Reserva):
    def __init__(self, nombre_del_pasajero, numero_de_vuelo, fecha, asiento, primeraclase):
        super().__init__(nombre_del_pasajero, numero_de_vuelo, fecha)
        self.asiento = asiento
        self.primeraclase = primeraclase

    def mostrar_detalle(self):
        padre = super().mostrar_detalle()
        return f"{padre}\nClase: Primera Clase\nAsiento: {self.asiento}\nServicio de Champagne: {self.primeraclase}"

reserva_economica = ReservaEconomica("Juan Pérez", "AA123", "2023-09-20", "Economica 23A")
reserva_business = ReservaBusiness("María García", "BA456", "2023-09-20", "vip 5B", "Sí")
reserva_primera_clase = ReservaPrimeraClase("Pedro Rodríguez", "CX789", "2023-09-20", "Primera Clase 1A", "Sí")

print("Detalle de la reserva económica:")
print(reserva_economica.mostrar_detalle())

print("\nDetalle de la reserva de business:")
print(reserva_business.mostrar_detalle())

print("\nDetalle de la reserva de primera clase:")
print(reserva_primera_clase.mostrar_detalle())
